File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    moves = {'U': (0, 1), 'D': (0, -1), 'R': (1, 0), 'L': (-1, 0)}
    visited = set()  # 간선(길) 기록
    x, y = 0, 0

    for d in dirs:
        dx, dy = moves[d]
        nx, ny = x + dx, y + dy
        
        # 경계 체크
        if -5 <= nx <= 5 and -5 <= ny <= 5:
            # 양방향 길 기록 (정렬된 튜플 형태)
            path = ((x, y), (nx, ny))
            reverse_path = ((nx, ny), (x, y))
            
            if path not in visited and reverse_path not in visited:
                visited.add(path)
                visited.add(reverse_path)
            
            # 위치 업데이트
            x, y = nx, ny

    # 서로 반대방향을 둘 다 기록했으니 길의 개수는 절반!
    return len(visited) // 2

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    const moves = {
        'U': [0, 1],
        'D': [0, -1],
        'R': [1, 0],
        'L': [-1, 0]
    };

    const visited = new Set();
    let x = 0, y = 0;

    for (let d of dirs) {
        const [dx, dy] = moves[d];
        const nx = x + dx;
        const ny = y + dy;

        // 경계 체크
        if (nx < -5 || nx > 5 || ny < -5 || ny > 5) {
            continue;
        }

        // 길 기록 (양방향 저장)
        const path = `${x},${y},${nx},${ny}`;
        const reversePath = `${nx},${ny},${x},${y}`;

        visited.add(path);
        visited.add(reversePath);

        // 위치 업데이트
        x = nx;
        y = ny;
    }

    // 길은 양방향으로 저장했으므로 2로 나눔
    return visited.size / 2;
}
